src/dataflow-types/types.rs

```diff
@@ -281,52 +281,54 @@ pub enum DataEncoding {
 
 impl DataEncoding {
     pub fn desc(&self, envelope: &Envelope) -> Result<RelationDesc, failure::Error> {
-        let full_desc = if let Envelope::Upsert(key_encoding) = envelope {
-            let key_desc = key_encoding.desc(&Envelope::None)?;
-            //rename key columns to "key" something if the encoding is not Avro
-            let key_desc = match key_encoding {
-                DataEncoding::Avro(_) => key_desc,
-                _ => RelationDesc::new(
-                    key_desc.typ().clone(),
-                    key_desc
-                        .iter_names()
-                        .enumerate()
-                        .map(|(i, _)| Some(format!("key{}", i))),
-                ),
-            };
-            let keys = key_desc
-                .iter_names()
-                .enumerate()
-                .map(|(i, _)| i)
-                .collect::<Vec<_>>();
-            key_desc.add_keys(keys)
-        } else {
-            RelationDesc::empty()
+        let desc = match envelope {
+            Envelope::Upsert(key_encoding) => {
+                let desc = key_encoding.desc(&Envelope::None)?;
+
+                // It doesn't make sense for the key to have keys.
+                assert!(desc.typ().keys.is_empty());
+
+                // Add the key columns as a key.
+                let key = (0..desc.arity()).collect();
+                let desc = desc.with_key(key);
+
+                // Rename key columns to "keyN" if the encoding is not Avro.
+                match key_encoding {
+                    DataEncoding::Avro(_) => desc,
+                    _ => {
+                        let names = (0..desc.arity()).map(|i| Some(format!("key{}", i)));
+                        desc.with_names(names)
+                    }
+                }
+            }
+            _ => RelationDesc::empty(),
```

Comment:
I can't tell what the before or the after are doing, so .. seems fine. Maybe a good time to document the method?

Reply:
Done.